src/lib/core/event.c

```diff
@@ -0,0 +1,336 @@
+ */
+static void
+event_trigger_delete(struct event_trigger *trigger)
+{
+	assert(trigger->func != NULL);
+	assert(trigger->event != NULL);
+	rlist_del(&trigger->link);
+	event_unref(trigger->event);
+	func_adapter_destroy(trigger->func);
+	TRASH(trigger);
+	free(trigger);
+}
+
+/**
+ * Increments event_trigger reference counter.
+ */
+static void
+event_trigger_ref(struct event_trigger *trigger)
+{
+	trigger->ref_count++;
+}
+
+/**
+ * Decrements event_trigger reference counter. The event_trigger is destroyed
+ * if the counter reaches zero.
+ */
+static void
+event_trigger_unref(struct event_trigger *trigger)
+{
+	assert(trigger != NULL);
+	assert(trigger->ref_count > 0);
+	trigger->ref_count--;
+	assert(trigger->ref_count > 0 || trigger->is_deleted);
+	if (trigger->ref_count == 0)
+		event_trigger_delete(trigger);
+}
+
+/**
+ * Allocates, initializes a new event and inserts it to event registry.
+ * Passed name will be copied.
+ */
+static struct event *
+event_new(const char *name, size_t name_len)
+{
+	assert(name != NULL);
+	assert(name_len > 0);
+	struct event *event = xmalloc(sizeof(*event) + name_len + 1);
+	rlist_create(&event->triggers);
+	event->ref_count = 0;
+	event->trigger_count = 0;
+	event->name = (char *)(event + 1);
+	strlcpy(event->name, name, name_len + 1);
+
+	struct mh_strnptr_t *h = event_registry;
+	name = event->name;
+	uint32_t name_hash = mh_strn_hash(name, name_len);
+	struct mh_strnptr_node_t node = {name, name_len, name_hash, event};
+	struct mh_strnptr_node_t prev;
+	struct mh_strnptr_node_t *prev_ptr = &prev;
+	mh_int_t pos = mh_strnptr_put(h, &node, &prev_ptr, NULL);
+	(void)pos;
+	assert(pos != mh_end(h));
+	assert(prev_ptr == NULL);
+
+	return event;
+}
+
+/**
+ * Destroys an event and removes it from event registry.
+ * Underlying trigger list is destroyed.
+ */
+void
+event_delete(struct event *event)
+{
+	assert(event != NULL);
+
+	struct mh_strnptr_t *h = event_registry;
+	const char *name = event->name;
+	size_t name_len = strlen(name);
+	mh_int_t i = mh_strnptr_find_str(h, name, name_len);
+	assert(i != mh_end(h));
+	assert(mh_strnptr_node(h, i)->val == event);
+	mh_strnptr_del(h, i, 0);
+
+	/*
+	 * If the only thing that holds the event is its trigger list,
+	 * the reference counter will reach zero when the list will be cleared
+	 * and the destructor will be called again. Let's reference the event
+	 * in order to prevent such situation.
+	 */
+	event_ref(event);
+	struct event_trigger *trigger, *tmp;
+	rlist_foreach_entry_safe(trigger, &event->triggers, link, tmp) {
+		/* Expected that the trigger is referenced only by event. */
+		assert(trigger->ref_count == 1);
+		event_trigger_delete(trigger);
+	}
+	TRASH(event);
+	free(event);
+}
+
+/**
+ * Finds internal event_trigger entity by name in the event.
+ * All arguments must not be NULL.
+ */
+static struct event_trigger *
+event_find_trigger_internal(struct event *event, const char *name)
+{
+	if (rlist_empty(&event->triggers))
+		return NULL;
```

Comment:
I thought that `rlist_foreach_empty` can lead to UB with an empty list because we will have a pointer to non-allocated memory in initialize statement of the `for` loop. But we call this function with empty lists everywhere, so I removed this check.